### PR TITLE
Partial damage cheat fix: node damages server side

### PR DESCRIPTION
**Pull request by nerzhul**
This PR changed, now we new protocol version we only calculate node damages server side

### CODE REVIEW

**Comment by nerzhul** [comment]
The remaining point to fix is https://github.com/minetest/minetest/pull/4981/files#diff-baa4d2d0ffafdec2b86baf065fba252bR2450 and it's not easy

File: src/game.cpp [hunk 1]
```diff
@@ -3535,7 +3535,6 @@ void Game::processClientEvents(CameraOrientation *cam, float *damage_flash)
 		} else if (event.type == CE_HUDADD) {
 			u32 id = event.hudadd.id;
 
-			LocalPlayer *player = client->getEnv().getLocalPlayer();
```
Comment (nerzhul) [inline src/game.cpp]: When i looked at this function i see we call this function here but it was already called in the parent scope.

File: src/localplayer.cpp [hunk 2]
```diff
@@ -313,7 +313,7 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
 
 			// The node to be sneaked on has to be walkable
 			node = map->getNodeNoEx(p, &is_valid_position);
-			if (!is_valid_position || nodemgr->get(node).walkable == false)
+			if (!is_valid_position || !nodemgr->get(node).walkable)
```
Comment (kwolekr) [inline src/localplayer.cpp]: Is this change actually necessary

**Comment by kwolekr** [comment]
It's hard to see from the diff what things have actually changed, since there are many lines simply moved around.

Do you care to give a complete explanation on how this PR fixes the issue?

File: src/content_sao.cpp [hunk 3]
```diff
@@ -969,6 +969,29 @@ void PlayerSAO::step(float dtime, bool send_recommended)
...
+		v3s16 p1 = floatToInt(m_base_position + v3f(0, BS*0.1, 0), BS);
+		MapNode n1 = m_env->getMap().getNodeNoEx(p1);
+		v3s16 p2 = floatToInt(m_base_position + v3f(0, BS*0.8, 0), BS);
+		MapNode n2 = m_env->getMap().getNodeNoEx(p2);
+		v3s16 p3 = floatToInt(m_base_position + v3f(0, BS*1.6, 0), BS);
+		MapNode n3 = m_env->getMap().getNodeNoEx(p3);
+
+		u32 damage_per_second = 0;
+		damage_per_second = MYMAX(damage_per_second,
+			((Server*) m_env->getGameDef())->ndef()->get(n1).damage_per_second);
```
Comment (kwolekr) [inline src/content_sao.cpp]: No, no, no.

Please think of encapsulation and separation of concerns here!

Why should a PlayerSAO ever need to know about the map, or an nodedef?  You're making a twisted web of object dependencies.  There should be a clear and distinct heirarchy.

If you find yourself needing to write code like "((Server*) m_env->getGameDef())->ndef()->get(n1)", you know you're doing it wrong.

This is not acceptable to me.

Comment (nerzhul) [inline src/content_sao.cpp]: I see your complain, now please propose a concrete solution to this problem, which is present everywhere in code base

File: src/localplayer.cpp (hunk 2 above)
Comment (nerzhul) [inline src/localplayer.cpp]: in fact no, i was only editing the file and see it, but i can remove it, to ensure only PR comment is covered

**Comment by nerzhul** [comment]
@kwolekr the issue is not completely solved at this moment. The solution is simple on the paper not in the facts for this problems.

Server should ignore TOSERVER_DAMAGE. This packet at this moment is send when: you are in lava or DPS node, and you fall.
With the current state of the PR:
* we don't send the packet anymore and it's ignored by server.
* Node DPS damages are calculated server side by PlayerSAO (object trigger environment damages for his position)
* we totally ignore falling damages and this should be repaired but it's not very simple as it call client side collisions functions which doesn't really exists server side

**Comment by nerzhul** [comment]
I worked on it, it's better now, but a speed desync can cause falling damages to bet forget by server and sometimes you take damages multiple times from a single fall

**Comment by sofar** [comment]
Wouls be nice if this could make it into the milestone!

**Comment by nerzhul** [comment]
@sofar the main problem is the fall detection which is very complicated... i think we can partially integrate it (node damages & breath damages) with the last protocol version, and differ the last part (collision damages) for next release as it's complicated to sync player with accurate. Are you okay with that ?

**Comment by sofar** [comment]
Yes, I have no issue with small incremental changes for the better ;)

**Comment by nerzhul** [comment]
Okay, then i will extract the interesting part for the current protocol version when i will rebase it

**Comment by nerzhul** [comment]
@sofar PR is now reduced and node damages are only sent for proto < 29 & calculated server side

**Comment by sofar** [comment]
This looks really, really trivial now.

I need to test this - tomorrow.

**Comment by sofar** [comment]
```
/home/sofar/git/minetest/src/content_sao.cpp: In member function ‘virtual void PlayerSAO::step(float, bool)’:
/home/sofar/git/minetest/src/content_sao.cpp:962:35: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
    s16 newhp = (damage_per_second > m_hp ? 0 : m_hp - damage_per_second);
                 ~~~~~~~~~~~~~~~~~~^~~~~~
```

**Comment by sofar** [comment]
Standing entirely in lava - not taking damage.

**Comment by sofar** [comment]
the breath bubbles also don't go away now.

**Comment by nerzhul** [comment]
okay i will look at this soon

**Comment by nerzhul** [comment]
@sofar oh, i just forgot to send the hp packet to client it's now fixed heh :) (i also rebased)

File: src/content_sao.cpp [hunk 4]
```diff
@@ -941,6 +941,30 @@ void PlayerSAO::step(float dtime, bool send_recommended)
...
+
+		u32 damage_per_second = 0;
+		damage_per_second = MYMAX(damage_per_second,
+			m_env->getGameDef()->ndef()->get(n1).damage_per_second);
+		damage_per_second = MYMAX(damage_per_second,
+			m_env->getGameDef()->ndef()->get(n2).damage_per_second);
+		damage_per_second = MYMAX(damage_per_second,
+			m_env->getGameDef()->ndef()->get(n3).damage_per_second);
+
+		if (damage_per_second != 0) {
```
Comment (SmallJoker) [inline src/content_sao.cpp]: `&& m_hp > 0`
No need to send further packets when the player already died.

Comment (nerzhul) [inline src/content_sao.cpp]: exact nice shot

**Comment by sofar** [comment]
:+1:

I also verified a new server with this patch works with clients that do not have this patch. Damage is slightly more laggy this way, which is expected.

**Comment by sofar** [comment]
Take the WIP out of the commit message? Or squash it.

**Comment by nerzhul** [comment]
it's done :) thanks for the review
